File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/maximum-number-of-non-overlapping-palindrome-substrings.cpp
class Solution {
public:
    int maxPalindromes(string s, int k) {
        int n = s.size();

        vector<vector<bool>> dp(
            n + 1, vector<bool>(n + 1, false)
        );

        for (int len = 1; len <= n; len++) {
            for (int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;

                if (len == 1) {
                    dp[i][j] = true;
                }
                else if (len == 2 && s[i] == s[j]) {
                    dp[i][j] = true;
                }
                else if (len >= 3) {
                    if (s[i] == s[j] && dp[i + 1][j - 1]) {
                        dp[i][j] = true;
                    }
                }
            }
        }

        vector<int> arr(n + 1, 0);

        for (int i = 1; i <= n; i++) {
            arr[i] = arr[i - 1];

            for (int j = 0; j < i; j++) {
                int len = i - j;

                if (len >= k && dp[j][i - 1]) {
                    arr[i] = max(arr[i], arr[j] + 1);
                }
            }
        }

        return arr[n];
    }
};

File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/maximum-number-of-non-overlapping-palindrome-substrings.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        n = len(s)

        dp = [[False for _ in range(n + 1)] for _ in range(n + 1)]

        for leng in range(1 , n + 1):
            for i in range(0 , n - leng + 1):
                j = i + leng - 1

                if leng == 1:
                    dp[i][j] = True
                
                elif leng == 2 and s[i] == s[j]:
                    dp[i][j] = True
                
                elif leng >= 3:
                    if s[i] == s[j] and dp[i + 1][j - 1]:
                        dp[i][j] = True

        arr = [0 for _ in range(n + 1)]

        for i in range(1 , n + 1):
            arr[i] = arr[i - 1]

            for j in range(i):
                lent = i - j

                if lent >= k and dp[j][i - 1] == True:
                    arr[i] = max(arr[i] , arr[j] + 1)
    
        return arr[n]